.cargo/config.toml

```diff
@@ -1,38 +1,9 @@
-[alias]
-b = "build"
-c = "check"
-d = "doc"
-t = "test"
-r = "run"
-
 [build]
 incremental = true
-# might be helpful but we don't enforce sccache installation atm
-# https://github.com/mozilla/sccache
-# rustc-wrapper = "sccache"
-
-# For details checkout https://jondot.medium.com/8-steps-for-troubleshooting-your-rust-build-times-2ffc965fd13e
-[target.x86_64-unknown-linux-gnu]
-# clang has been listed as prerequisite in the doc
-linker = "clang"
```

Comment:
What is the implication of removing this? I believe it was added to reduce linking time. If we still want to remove it, the dependencies in both README and Dockerfiles (and Makefile) should be updated.

Reply:
Do we own linking time? Or is it owned by the developer?

I'd argue that we don't own any cargo config:
- Devs will have their own linkers set up in in `~/.cargo/config.toml`. E.g if they've install `mold`
- Fetching with CLI, cargo target directory etc, that's very developer-owned

The gray area here is `--cfg=tokio_unstable`, but I think that's a minor feature, and if users want to include it, we should provide instructions.

Note if we supported a devcontainer, we _would_ then own more of the developer experience - linking building etc, but at the moment we're a mess of both (I guess this is why I'm spending so much time here).

We don't really want to be messing with build config too much, because it ends up being pain - we're not build experts, we're rust experts.
One sample is the vectorised instruction support which we've mentioned earlier. This is another excellent example
- #2104 Added this https://github.com/ChainSafe/forest/pull/2104/files#diff-9a4f3e4537ebb7474452d131b0d969d89a51286f4269aac5ef268e712be17268R19
- #2200 Removes the link arg that actually speeds up linking, leaving us with technical debt in our build args 'linker - "clang"' that we're not sure works

BTW this arg does nothing on my machine, you actually need a linker if you want a speedup:
```
  './do-link.bash -C link-arg=-fuse-ld=mold' ran
    1.02 ± 0.11 times faster than './do-link.bash -C linker=clang -C link-arg=-fuse-ld=mold'
    1.21 ± 0.08 times faster than './do-link.bash -C linker=clang -C link-arg=-fuse-ld=lld'
    1.22 ± 0.08 times faster than './do-link.bash -C link-arg=-fuse-ld=lld'
    4.81 ± 0.32 times faster than './do-link.bash -C linker=clang'
    4.95 ± 0.42 times faster than './do-link.bash'
```
See benchmarking at https://github.com/ChainSafe/forest/commit/54fb5e8c29f3e28c3414cf5e63f2acddfec65b39

I'm generally all for build-time tuning, and developer experience, but I think the way we handle it at the moment is a bit haphazard, and we end up dropping balls outside of the actual source code

Reply:
I second this motion. I'd have more confidence in the correctness of Forest if we used the default configuration.

Reply:
I mean, I overall agree that using defaults is the way to go. That being said if a couple of lines of configuration that are rarely changed bring 10-20% faster builds for thousands of workflow runs in the future - IMO it's worth having it as long as it is not really exotic piece of technology that changes a lot.

Reply:
> we do own the CI
Yeah that's an important point I missed.

# Current suggestion
- remove `linker = "clang"`, it doesn't do anything
- @elmattic is doing profiling on the avx2 stuff. I'm partial to removing it in any case

# Looking forward
cc #2826 

I think we should split responsibilites:
- forest only deals with writing and shipping minimal, correct rust (and javascript???) code
  (we bin our custom build config)

Then I'd lean towards a CI container with:
- custom build config - sccache, custom linkers, (maybe) cached deps
- cached installs of tooling and dev-deps, so we're never stuck on apt dependencies etc
- a nightly job that ensures our CI container is taking the latest tooling
- support for developers using CI containers as a devcontainer

I think we should then remove everything that we can here. Even the tokio console support should just be some `#[cfg(..)]` gated code, there's no need for us to touch that in-repo

There's some outstanding thinking on "where do locks" live - I don't really want to have to coordinate lockfiles across two repos, but maybe if we see the CI container as more of a "caching layer" then we can help ourselves a little bit

Reply:
P.S

# What is performance engineering?
I think vectorizing our code/dependencies is a matter of "performance engineering", maybe that should live in our CI repo - we compile and build/distribute an optimised/stripped binary. But where our product is now, I think we're a long way off - we don't have benchmarking infra, and our code quality/organisation isn't disciplined enough to be worried about e.g vectorisation - we should focus on a portable, correct and ergonomic implementation first and foremost

Reply:
@aatifsyed Regarding "performance engineering", to give you a bit of context. We did quite a bit of benchmarking some time ago. Before that, the time of importing and syncing a snapshot was counted in many hours, right now it's below 1 hour on a regular developer machine. Most of the gains came from non-flag optimisations (changing the DB, checkpoints) but even then I feel uneasy when something may possibly deteriorate this. We have a benchmarking script (and its more advanced version coming up in https://github.com/ChainSafe/forest/pull/2367).

Reply:
I'm fine with removing `clang` given the benchmarks you've submitted. If we can remove it from the dependencies, let's do it as well.

Reply:
Removing `avx2` - I'm fine as long as the performance hit is not big, below 5%. Hopefully @elmattic will provide numbers.

Reply:
As for other points - no strong opinion. I didn't have issues with the current tooling on either Linux Fedora or MacOS. NPM is sometimes down, but so are Github and Ubuntu repositories. I'm good with minimising the impact of such outages but the only ones that are really annoying me are Github outages. Little we can do about it.

Reply:
I have spoken. :mountain:

Reply:
# Impact of `avx2`
I've not got enough hard drive space to use mainnet snapshots, so I've done it with calibnet instead.
## Methodology
I built the code with and without avx here: https://github.com/ChainSafe/forest/blob/c5b75126b84b1c5af98057187fe43a3f9a96cdff/build.bash
I ran the benchmark script here https://github.com/ChainSafe/forest/blob/c5b75126b84b1c5af98057187fe43a3f9a96cdff/benchmark.bash
I didn't have enough hard drive space to run against mainnet snapshots, so I used calibnet
## Importing
```
```
## Validating
```
```

Reply:
I have on my side:

```
Network: mainnet
Snapshot used: 2866560_2023_05_17T06_00_00Z.car
OS: Fedora 37
CPU: Intel Core i3-6300
```

| Command | Import Snapshot | Validation of 30 heights
|:---|---:|---:|
| `avx2` | 14m22s | 2m16s |
| `no_avx2` | 14m13s | 2m12s |

Reply:
If you can confirm this on a `mainnet` snapshot as well.

Reply:
I would say we can leave those instructions away, and re-assess performance later when needed.

Reply:
Thanks for assessing this, LGTM.